fix(ivy): normalize summary and factory shim files paths

alan-agius4 opened the pull request:
At the moment, paths stored in `maps` are not normalized and in Windows is causing files not to be found when enabling factory shimming.

For example, the map contents will be
```
Map {
  'C:\\git\\cli-repos\\ng-factory-shims\\index.ngfactory.ts' => 'C:\\git\\cli-repos\\ng-factory-shims\\index.ts' }
```

However, ts compiler normalized the paths and is causing;
```
error TS6053: File 'C:/git/cli-repos/ng-factory-shims/index.ngfactory.ts' not found.
error TS6053: File 'C:/git/cli-repos/ng-factory-shims/index.ngsummary.ts' not found.
```

The changes normalized the paths that are stored within the factory and summary maps.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

alxhub commented on packages/compiler-cli/src/ngtsc/shims/BUILD.bazel:
```diff
@@ -13,6 +13,7 @@ ts_library(
         "//packages/compiler",
         "//packages/compiler-cli/src/ngtsc/util",
         "@ngdeps//@types/node",
+        "@ngdeps//canonical-path",
```
This would have to be added to compiler-cli's package.json as well.

I'll have to think about this for a bit. We currently have only a single dependency (TypeScript itself) in compiler-cli. I don't know if it's worth incurring an external dependency purely for path normalization, which seems like something we could do ourselves in a more type-safe way.

alan-agius4 replied:
This would work as well
https://github.com/alan-agius4/angular/blob/b69e33e39aa99f2e02e51e8cf2584826bdceae8e/packages/compiler-cli/src/transformers/program.ts#L967

alan-agius4 replied:
@alxhub, updated the Pr

alan-agius4 committed:
fix(ivy): normalize summary and factory shim files paths

At the moment, paths stored in `maps` are not normalized and in Windows is causing files not to be found when enabling factory shimming.

For example, the map contents will be
```
Map {
  'C:\\git\\cli-repos\\ng-factory-shims\\index.ngfactory.ts' => 'C:\\git\\cli-repos\\ng-factory-shims\\index.ts' }
```

However, ts compiler normalized the paths and is causing;
```
error TS6053: File 'C:/git/cli-repos/ng-factory-shims/index.ngfactory.ts' not found.
error TS6053: File 'C:/git/cli-repos/ng-factory-shims/index.ngsummary.ts' not found.
```

The changes normalized the paths that are stored within the factory and summary maps.

alxhub commented on packages/compiler-cli/src/ngtsc/util/src/path.ts:
```diff
@@ -26,3 +26,7 @@ export function relativePathBetween(from: string, to: string): string|null {
   return relative;
 }
+
+export function normalizeSeparators(path: string): string {
```
Please add a `TODO` here to limit normalization when running under an OS which doesn't require it.

alxhub reviewed (approved):
Other than the one comment, LGTM!

alan-agius4 committed:
fixup! fix(ivy): normalize summary and factory shim files paths

gkalpak commented on packages/compiler-cli/src/ngtsc/shims/src/factory_generator.ts:
```diff
@@ -88,7 +89,8 @@ export class FactoryGenerator implements ShimGenerator {
     const map = new Map<string, string>();
-    files.filter(sourceFile => isNonDeclarationTsPath(sourceFile))
+    files.map(sourceFile => normalizeSeparators(sourceFile))
+        .filter(sourceFile => isNonDeclarationTsPath(sourceFile))
```
Nit: Why not normalize after filtering :grin:
\#MicroOptimizing

alan-agius4 replied:
loool, done

alan-agius4 committed:
fixup! fix(ivy): normalize summary and factory shim files paths

gkalpak commented:
@caretaker: Needs g3sync presubmit.

AndrewKushnir commented:
[Presubmit](http://test/OCL:229385719:BASE:229385750:1547573814387:8b7e9cb9)

AndrewKushnir commented:
@alan-agius4 I'm receiving conflicts while trying to merge this with merge script. It looks like it conflicts with patch branch. Could you please mark this PR as "PR target: master-only" and create a new one to target patch branch? Thank you.

AndrewKushnir commented:
Discussed with Alan via Slack. We've changed PR target to "master-only" for now, to merge this PR to master and Alan will create a new PR to target patch branch. Thank you.
